### PR TITLE
[Product Block Editor]: move Modal editor out of the description block edit component

### PR DESCRIPTION
**Pull request by retrofox**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This Pull Request (PR) significantly improves how our application handles editor components. It separates the Modal editor component from the Description block edit component. The primary effect of this change is that it prevents both editor instances from being rendered simultaneously in the same context. As a result, the application will now render only one editor at a time, enhancing performance and user experience.

#### Enhancements in Modal Editor Usage

Key additions in this PR include several actions integrated into the `woo/product-editor-ui` store. These actions are designed to manage the Modal editor's content. One of the actions acts like a signal that detects if the user has modified the content in the Modal editor. This functionality is essential for synchronizing content across different components of our application.
For example, it allows for seamless content integration from the Modal editor into the preview component of the Description block.

Here are the specific changes:
- The Modal editor's blocks are now stored within the product editor UI store. This centralizes the management of these blocks and simplifies their accessibility across the application.
- We have eliminated the practice of storing block instances from the iFrame editor in the local state. These instances are stored globally, aligning with our application's data management strategy.
- The Modal editor component has been relocated to be part of the regular product editor. This repositioning promotes a more cohesive user interface.

These changes aim to improve the maintainability and scalability of our editor components, ultimately enhancing the overall user experience.

Closes # .

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Go to the block editor
2. Edit the product description as usual
3. Additionally, you can use the Redux tool to see how the data is propagated through the app workflow.

https://github.com/woocommerce/woocommerce/assets/77539/35491aac-c2fa-451e-950e-a0745d23e30a

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: packages/js/product-editor/src/components/iframe-editor/iframe-editor.tsx [hunk 1]
```diff
@@ -42,20 +43,31 @@ type IframeEditorProps = {
...
+	// Pick the blocks from the store.
+	const blocks: BlockInstance[] = useSelect( ( select ) => {
+		return select( productEditorUiStore ).getModalEditorBlocks();
+	}, [] );
+
+	const {
+		setModalEditorBlocks: setBlocks,
+		closeModalEditor,
+		setModalEditorContentHasChanged,
+	} = useDispatch( productEditorUiStore );
```
Comment (louwie17) [inline packages/js/product-editor/src/components/iframe-editor/iframe-editor.tsx]: This feels very specific to the ModalEditor component, would this not be better handled within the `ModalEditor` component? or in the BlockEditor component where the ModalEditor is being used?

The IFrameEditor is kind of a stand alone thing, and should not need to worry/care about the context of its parent ( the modal editor ).

**Review by louwie17 — CHANGES_REQUESTED** [review]
Nice!! Moving this outside of the nested editor context makes sense.
Makes me wonder if we should actual use a portal for this? But for now this makes sense.

I did leave one suggestion in line, I am not a huge fan that the IFrameEditor has to be aware of the Modal editor store context. I think this should be handled up stream as part of the callback function. Either in the ModalEditor component or the BlockEditor component that handles the showing/hiding of the modal.

File: packages/js/product-editor/src/components/iframe-editor/iframe-editor.tsx (hunk 1 above)
Comment (retrofox) [reply packages/js/product-editor/src/components/iframe-editor/iframe-editor.tsx]: nice suggestion. Updating...

Comment (retrofox) [reply packages/js/product-editor/src/components/iframe-editor/iframe-editor.tsx]: Addresses https://github.com/woocommerce/woocommerce/pull/41878/commits/c44324dea4b5c38ca350f1399080e1ff707b7944

**Comment by retrofox** [comment]
> Nice!! Moving this outside of the nested editor context makes sense. Makes me wonder if we should actual use a portal for this? But for now this makes sense.

Using a portal could make sense. Let me create follow-up tasks:

* Consider to move the place where the Modal editor readers by using a portal 
* Currently, the iFrame editor accepts an `initalBlocks` property. We can probably discard it since the blocks can be defined through the store.
* Consider removing the debounce handling. It seems a few race conditions with the current implementation were addressed by using the debouncing technique. After these changes, we may discard it.

File: packages/js/product-editor/src/components/iframe-editor/iframe-editor.tsx [hunk 2]
```diff
@@ -42,20 +43,28 @@ type IframeEditorProps = {
...
 	const [ temporalBlocks, setTemporalBlocks ] =
 		useState< BlockInstance[] >( initialBlocks );
+
+	// Pick the blocks from the store.
+	const blocks: BlockInstance[] = useSelect( ( select ) => {
+		return select( productEditorUiStore ).getModalEditorBlocks();
+	}, [] );
+
+	const { setModalEditorBlocks: setBlocks, setModalEditorContentHasChanged } =
+		useDispatch( productEditorUiStore );
```
Comment (louwie17) [inline packages/js/product-editor/src/components/iframe-editor/iframe-editor.tsx:63]: Is there a specific reason you still kept this as part of the iframe editor?
I was kinda thinking we could trigger these functions from the `onChange` menu in the `ModalEditor` component.

Likewise for the `blocks` I was thinking we could use the `initialBlocks`, or just pass in the `blocks` array so there is no intermediary state.

**Review by louwie17 — CHANGES_REQUESTED** [review]
Thanks for updating @retrofox, I was envisioning a bit more of it, so that the IframeEditor can purely rely on its callback functions, while the modal editor relies on the data store.
I am mostly pushing this, as I could see the IFrameEditor being used outside of the Modal editor, in which case the modal editor data store seems irrelevant.

If we think that won't be the case, I would prefer to just merge the IFrameEditor and the ModalEditor components together, and avoid the unnecessary layer.

File: packages/js/product-editor/src/components/iframe-editor/iframe-editor.tsx (hunk 2 above)
Comment (retrofox) [reply packages/js/product-editor/src/components/iframe-editor/iframe-editor.tsx:63]: I'd like to think blocks belong to the block (Iframe) editor, while the modal visibility belongs to the Modal editor.

I'd remove also all props from these components that are provided by the store.

**Comment by retrofox** [comment]
> ...I was envisioning a bit more of it, so that the IframeEditor can purely rely on its callback functions, while the modal editor relies on the data store.

I'd like to remove all properties of these components if it's possible. For instance, we don't need to pass the `initialBlocks` prop because the iFrame editor can pick them up from the store. Passing down these props between parent and child components doesn't look good IMO. Same with the rest of the props.

> I am mostly pushing this, as I could see the IFrameEditor being used outside of the Modal editor, in which case the modal editor data store seems irrelevant.

Yeah, the Modal editor doesn't need to take care of the data about the editor but, only its visibility.

> If we think that won't happen, I would prefer to merge the IFrameEditor and ModalEditor components, avoiding the unnecessary layer.

I imagine some scenarios where we can use the iFrame editor independently of the Modal editor component. Anyway, we could merge both parts into a single one. If so, I suggest doing that in a follow-up.

**Review by louwie17 — APPROVED** [review]
Thanks for the updates @retrofox, this seems like a good solution. Also the data store as discussed in Slack is fine as an intermediary solution for now.

**Pull request merged**